### PR TITLE
GEODE-4399 dotnet callback optional

### PR DESCRIPTION
**Pull request by moleske**
If you didn't review #303 then go review that first.  After that's merge can rebase and try this one.  Opening to start the conversation (before I forget context here).

### CODE REVIEW

File: clicache/integration-test/CMakeLists.txt [hunk 1]
```diff
@@ -103,10 +103,9 @@ set_property(TEST OverflowTestsN PROPERTY LABELS OMITTED)
 set_property(TEST ThinClientDeltaTestN PROPERTY LABELS OMITTED)
 set_property(TEST ThinClientDurableTestsN PROPERTY LABELS OMITTED)
 set_property(TEST ThinClientHARegionTestsN PROPERTY LABELS OMITTED)
-set_property(TEST ThinClientRegionTestsN PROPERTY LABELS OMITTED)
 set_property(TEST ThinClientSecurityAuthTestsMUN PROPERTY LABELS OMITTED)
 set_property(TEST ThinClientSecurityAuthTestsN PROPERTY LABELS OMITTED)
-set_property(TEST ThinClientSecurityAuthTestsMUN PROPERTY LABELS OMITTED)
+set_property(TEST RegionFailingTests PROPERTY LABELS OMITTED)
```
Comment (mcmellawatt) [inline clicache/integration-test/CMakeLists.txt]: Did we introduce new failing tests and mark them as omitted as part of this changeset?

File: clicache/integration-test/DistOpsStepsN.cs [hunk 2]
```diff
@@ -121,7 +121,7 @@ public abstract class DistOpsSteps : UnitTests
           }
           else
           {
-            string cVal = region.Get(cKey, true).ToString();
+            string cVal = region.Get(cKey).ToString();
```
Comment (mcmellawatt) [inline clicache/integration-test/DistOpsStepsN.cs]: var

File: clicache/integration-test/DistOpsStepsN.cs [hunk 3]
```diff
@@ -165,7 +165,7 @@ protected virtual void VerifyCreated(string regionName, string key)
 
     protected virtual object GetEntry(string regionName, string key)
     {
-      object val = CacheHelper.GetVerifyRegion<string, string>(regionName).Get(key, null);
+      object val = CacheHelper.GetVerifyRegion<string, string>(regionName).Get(key);
```
Comment (mcmellawatt) [inline clicache/integration-test/DistOpsStepsN.cs]: var

File: clicache/integration-test2/RegionSSLTest.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+{
+  [Trait("Category", "Integration")]
+  public class RegionSSLTest : IDisposable
+  {
+    private readonly Cache _cacheOne;
+    private readonly GeodeServer _geodeServer;
+
+    public RegionSSLTest()
+    {
+      var pathvar = Environment.GetEnvironmentVariable("PATH");
```
Comment (mcmellawatt) [inline clicache/integration-test2/RegionSSLTest.cs]: I like more test coverage, but what does this test have to do with the ticket?  Maybe I am missing some relationship between SSL and optional callback parameters.

Comment (moleske) [reply clicache/integration-test2/RegionSSLTest.cs]: We didn't get a callback test with SSL working in the `RegionFailingTests.cs` so we added one.

File: clicache/integration-test/CMakeLists.txt (hunk 1 above)
Comment (moleske) [reply clicache/integration-test/CMakeLists.txt]: If you look at the test names, you should be able to see them line up with what use to be in the `ThinClientRegionTestsN` file.

Comment (mcmellawatt) [reply clicache/integration-test/CMakeLists.txt]: Got it, just making sure.  I agree this is an improvement.

File: clicache/integration-test2/RegionSSLTest.cs (hunk 4 above)
Comment (pivotal-jbarrett) [reply clicache/integration-test2/RegionSSLTest.cs]: I also fail to see why it is included in this PR.

File: clicache/integration-test/BuiltinCacheableWrappersN.cs [hunk 5]
```diff
@@ -270,6 +271,7 @@ public static void RegisterBuiltinsJavaHashCode(long dateTime)
         typeof(CacheableHugeUnicodeStringWrapper), CacheableHugeUnicodeStringWrapper.Create);
 
       //need to register pdx type
+      Assert.NotNull(CacheHelper.DCache);
```
Comment (pivotal-jbarrett) [inline clicache/integration-test/BuiltinCacheableWrappersN.cs]: Why are we asserting framework variables? Asserts should be reserved for things that are part of the test qualifications. Exceptions should be thrown for things that fail as a result of tooling issues.

File: clicache/integration-test/DistOpsStepsN.cs [hunk 6]
```diff
@@ -2328,7 +2317,7 @@ public virtual void IdictionaryRegionOperations(String RegionName)
       Util.Log("Idictionary_Item_Add_Get_Set_Remove_Step 4 complete.");
 
       // Chk with old api that element that is added by new api exists.
-      Assert.AreEqual(type.GetTypeItem<TValue>(ValueId, 0), region.Get(type.GetTypeItem<TKey>(KeyId, 0), null));
+      Assert.AreEqual(type.GetTypeItem<TValue>(ValueId, 0), region.Get(type.GetTypeItem<TKey>(KeyId, 0)));
```
Comment (pivotal-jbarrett) [inline clicache/integration-test/DistOpsStepsN.cs]: Do we still have tests that assert operations complete successfully when `null` is the callback value?

File: clicache/integration-test2/Config.cs.in [hunk 7]
```diff
@@ -19,8 +19,18 @@
...
 {
-  static public string GeodeGfsh
+  public static string GeodeGfsh
   {
     get { return @"@Geode_gfsh_EXECUTABLE@"; }
   }
+
+  public static string CryptoImplPath
+  {
+    get { return @"/../../../cryptoimpl/@CMAKE_BUILD_TYPE@"; }
```
Comment (pivotal-jbarrett) [inline clicache/integration-test2/Config.cs.in]: NO! That is isn't how you do this in CMake. Setup a variable in CMake that is something like `CRYPTO_IMPL_PATH` and changes this source to `@"@CRYPTO_IMPL_PATH@"`.

File: clicache/integration-test2/Config.cs.in [hunk 8]
```diff
@@ -19,8 +19,18 @@
...
+  }
+
+  public static string OpenSSLPath
+  {
+    get { return @"/../../../dependencies/openssl/openssl-extern-prefix/@CMAKE_BUILD_TYPE@/bin"; }
```
Comment (pivotal-jbarrett) [inline clicache/integration-test2/Config.cs.in]: Same as above.

File: clicache/src/IRegion.hpp [hunk 9]
```diff
@@ -641,6 +635,17 @@ namespace Apache
 	      /// and propagate the operation to server.
           virtual void Clear();
 
+          /// <summary>
+          /// remove all entries in the region.	
+          /// For local region instance - remove all entries in the local region.
+          /// For distributed region instance - remove all entries in the local region,
+	      /// and propagate the operation to server.
```
Comment (pivotal-jbarrett) [inline clicache/src/IRegion.hpp]: Format

File: tests/cli/NewFwkLib/PdxTest/PdxTests.cs [hunk 10]
```diff
@@ -1582,595 +1582,8 @@ protected void verifySize(IRegion<TKey, TVal> m_region, int expectedSize)
                 throw new Exception("Expected size of " + m_region.FullPath + " to be " +
                    expectedSize + ", but it is " + size);
             }
-            // rjk:need to comment this block
```
Comment (pivotal-jbarrett) [inline tests/cli/NewFwkLib/PdxTest/PdxTests.cs]: Making no other changes in this file other than to delete commented out code seems out of scope. Please justify.

File: clicache/integration-test/BuiltinCacheableWrappersN.cs (hunk 5 above)
Comment (moleske) [reply clicache/integration-test/BuiltinCacheableWrappersN.cs]: This was us trying to figure out why some of the tests marked `omitted` were failing.  We found out that some of those tests were failing because `DCache` was null.  Since it was non-obvious that this was the cause we left this statement in to help in future for fixing the tests still `omitted`

File: clicache/integration-test2/RegionSSLTest.cs (hunk 4 above)
Comment (echobravopapa) [reply clicache/integration-test2/RegionSSLTest.cs]: We encountered a test that purported to cover the callback usage and SSL, but was in the old framework and riddled with decayed code so we took the opportunity to create a simple SSL test for the `clicache` while in the vicinity...

File: clicache/integration-test/DistOpsStepsN.cs (hunk 6 above)
Comment (echobravopapa) [reply clicache/integration-test/DistOpsStepsN.cs]: yes

File: tests/cli/NewFwkLib/PdxTest/PdxTests.cs (hunk 10 above)
Comment (echobravopapa) [reply tests/cli/NewFwkLib/PdxTest/PdxTests.cs]: clean up while hunting for tests using the callback

File: clicache/integration-test2/Config.cs.in (hunk 7 above)
Comment (echobravopapa) [reply clicache/integration-test2/Config.cs.in]: will have a go at that

File: clicache/integration-test2/Config.cs.in (hunk 8 above)
Comment (echobravopapa) [reply clicache/integration-test2/Config.cs.in]: ditto

File: clicache/integration-test2/RegionSSLTest.cs (hunk 4 above)
Comment (pivotal-jbarrett) [reply clicache/integration-test2/RegionSSLTest.cs]: meh.... ok

File: clicache/integration-test/BuiltinCacheableWrappersN.cs (hunk 5 above)
Comment (pivotal-jbarrett) [reply clicache/integration-test/BuiltinCacheableWrappersN.cs]: :/
